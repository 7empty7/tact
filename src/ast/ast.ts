import { Interval as RawInterval, Node as RawNode } from 'ohm-js';

export class ASTRef {
    readonly #interval: RawInterval;

    constructor(interval: RawInterval) {
        this.#interval = interval;
    }

    get contents() {
        return this.#interval.contents;
    }

    get interval() {
        return this.#interval;
    }
}

export type ASTPrimitive = {
    kind: 'primitive',
    id: number,
    name: string,
    ref: ASTRef
}

//
// Values
//

export type ASTNumber = {
    kind: 'number',
    id: number,
    value: bigint,
    ref: ASTRef
}

export type ASTID = {
    kind: 'id',
    id: number,
    value: string,
    ref: ASTRef
}

export type ASTBoolean = {
    kind: 'boolean',
    id: number,
    value: boolean,
    ref: ASTRef
}

export type ASTNull = {
    kind: 'null',
    id: number,
    ref: ASTRef
}

//
// Types
//

export type ASTTypeRef = {
    kind: 'type_ref',
    id: number,
    name: string,
    optional: boolean,
    ref: ASTRef
}

//
// Expressions
//

export type ASTOpBinary = {
    kind: 'op_binary',
    id: number,
    op: '+' | '-' | '*' | '/' | '!=' | '>' | '<' | '>=' | '<=' | '==' | '&&' | '||',
    left: ASTExpression,
    right: ASTExpression,
    ref: ASTRef
}

export type ASTOpUnary = {
    kind: 'op_unary',
    id: number,
    op: '+' | '-' | '!' | '!!',
    right: ASTExpression,
    ref: ASTRef
}

export type ASTOpField = {
    kind: 'op_field'
    id: number,
    src: ASTExpression,
    name: string,
    ref: ASTRef
}

export type ASTOpCall = {
    kind: 'op_call'
    id: number,
    src: ASTExpression,
    name: string,
    args: ASTExpression[],
    ref: ASTRef
}

export type ASTOpCallStatic = {
    kind: 'op_static_call'
    id: number,
    name: string,
    args: ASTExpression[],
    ref: ASTRef
}

export type ASTOpNew = {
    kind: 'op_new'
    id: number,
    type: string,
    args: ASTNewParameter[],
    ref: ASTRef
}

export type ASTNewParameter = {
    kind: 'new_parameter'
    id: number,
    name: string,
    exp: ASTExpression,
    ref: ASTRef
}

//
// Program
//

export type ASTProgram = {
    kind: 'program',
    id: number,
    entries: (ASTStruct | ASTContract | ASTPrimitive | ASTFunction | ASTNativeFunction)[]
}

export type ASTStruct = {
    kind: 'def_struct',
    id: number,
    name: string
    fields: ASTField[],
    ref: ASTRef
}

export type ASTField = {
    kind: 'def_field',
    id: number,
    name: string,
    type: ASTTypeRef,
    ref: ASTRef
}

export type ASTContract = {
    kind: 'def_contract',
    id: number,
    name: string,
    declarations: (ASTField | ASTFunction)[],
    ref: ASTRef
}

export type ASTArgument = {
    kind: 'def_argument',
    id: number,
    name: string,
    type: ASTTypeRef,
    ref: ASTRef
}

export type ASTFunctionAttribute = { type: 'public' } | { type: 'get' };

export type ASTFunction = {
    kind: 'def_function',
    id: number,
    attribute: ASTFunctionAttribute[],
    name: string,
    return: ASTTypeRef | null,
    args: ASTArgument[],
    statements: ASTStatement[],
    ref: ASTRef
}

export type ASTNativeFunction = {
    kind: 'def_native_function',
    id: number,
    name: string,
    nativeName: string,
    return: ASTTypeRef | null,
    args: ASTArgument[],
    ref: ASTRef
}

//
// Statements
//

export type ASTStatementLet = {
    kind: 'statement_let',
    id: number,
    name: string,
    type: ASTTypeRef,
    expression: ASTExpression,
    ref: ASTRef
}

export type ASTStatementReturn = {
    kind: 'statement_return',
    id: number,
    expression: ASTExpression,
    ref: ASTRef
}

export type ASTStatementCall = {
    kind: 'statement_call',
    id: number,
    expression: ASTOpCall | ASTOpCallStatic,
    ref: ASTRef
}

export type ASTSTatementAssign = {
    kind: 'statement_assign',
    id: number,
    path: string[],
    expression: ASTExpression,
    ref: ASTRef
}

//
// Unions
//

export type ASTStatement = ASTStatementLet | ASTStatementReturn | ASTStatementCall | ASTSTatementAssign;
export type ASTExpression = ASTOpBinary | ASTOpUnary | ASTOpField | ASTNumber | ASTID | ASTBoolean | ASTOpCall | ASTOpCallStatic | ASTOpNew;
export type ASTNode = ASTExpression | ASTProgram | ASTStruct | ASTField | ASTContract | ASTArgument | ASTFunction | ASTOpCall | ASTStatementLet | ASTStatementReturn | ASTProgram | ASTPrimitive | ASTOpCallStatic | ASTStatementCall | ASTNativeFunction | ASTSTatementAssign | ASTOpNew | ASTNewParameter | ASTTypeRef;
export type ASTType = ASTPrimitive | ASTStruct | ASTContract;

export function isStatement(src: ASTNode): src is ASTStatement {
    return src.kind === 'statement_let' || src.kind === 'statement_return' || src.kind === 'statement_call' || src.kind === 'statement_assign';
}

export function isExpression(src: ASTNode): src is ASTExpression {
    return src.kind === 'op_binary' || src.kind === 'op_unary' || src.kind === 'op_field' || src.kind === 'number' || src.kind === 'id' || src.kind === 'boolean' || src.kind === 'op_new' || src.kind === 'op_call' || src.kind === 'op_static_call';
}


type DistributiveOmit<T, K extends keyof any> = T extends any
    ? Omit<T, K>
    : never;
let nextId = 1;
export function createNode(src: DistributiveOmit<ASTNode, 'id'>): ASTNode {
    return Object.freeze(Object.assign({ id: nextId++ }, src));
}

export function createRef(s: RawNode, ...extra: RawNode[]): ASTRef {
    let i = s.source;
    if (extra.length > 0) {
        i = i.coverageWith(...extra.map((e) => e.source));
    }
    return new ASTRef(i);
}

export function throwError(message: string, ref: ASTRef): never {
    throw new Error(ref.interval.getLineAndColumnMessage() + message);
}