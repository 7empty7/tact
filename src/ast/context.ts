import { parse } from "../grammar/grammar";
import { ASTNode, ASTType } from "./ast";

export type VariableRef = {
    name: string,
    node: ASTNode
};

export class CompilerContext {

    static fromSources(sources: string[]) {
        let asts = sources.map(source => parse(source));
        let ctx = new CompilerContext();
        for (let a of asts) {
            for (let e of a.entries) {
                if (e.kind === 'def_struct') {
                    ctx = ctx.addASTType(e);
                } else if (e.kind === 'def_contract') {
                    ctx = ctx.addASTType(e);
                } else if (e.kind === 'primitive') {
                    ctx = ctx.addASTType(e);
                }
            }
        }
        return ctx;
    }

    readonly astTypes: { [key: string]: ASTType };
    readonly variables: { [key: string]: VariableRef };
    readonly shared: { [key: symbol]: any } = {};

    constructor(astTypes: { [key: string]: ASTType } = {}, variables: { [key: string]: VariableRef } = {}, shared: { [key: symbol]: any } = {}) {
        this.astTypes = astTypes;
        this.variables = variables;
        this.shared = shared;
        Object.freeze(this.astTypes);
        Object.freeze(this.shared);
        Object.freeze(this.variables);
        Object.freeze(this);
    }

    addASTType = (ref: ASTType) => {
        if (this.astTypes[ref.name]) {
            throw Error('Type already exists');
        }
        return new CompilerContext({ ...this.astTypes, [ref.name]: ref }, this.variables, this.shared);
    }

    addVariable = (ref: VariableRef) => {
        if (this.variables[ref.name]) {
            throw Error('Variable already exists');
        }
        return new CompilerContext(this.astTypes, { ...this.variables, [ref.name]: ref }, this.shared);
    }

    addShared = <T>(store: symbol, key: string | number, value: T) => {
        let sh: { [key: string]: T } = {};
        if (this.shared[store]) {
            sh = { ...this.shared[store] };
        }
        sh[key] = value;
        return new CompilerContext(this.astTypes, this.variables, { ...this.shared, [store]: sh });
    }
}

export function createContextStore<T>() {
    let symbol = Symbol();
    return {
        get(ctx: CompilerContext, key: string | number) {
            if (!ctx.shared[symbol]) {
                return null;
            }
            let m = ctx.shared[symbol] as { [key: string | number]: T };
            if (m[key]) {
                return m[key];
            } else {
                return null;
            }
        },
        set(ctx: CompilerContext, key: string | number, v: T) {
            return ctx.addShared(symbol, key, v);
        }
    }
}