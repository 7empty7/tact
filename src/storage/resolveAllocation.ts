import { CompilerContext, createContextStore } from "../ast/context";
import { getAllTypes, getType } from "../types/resolveTypeDescriptors";
import { FieldDescription, TypeDescription, TypeRef } from "../types/types";
import { topologicalSort } from "../utils";
import { crc32 } from "../utils/crc32";
import { StorageAllocation, StorageCell, StorageField } from "./StorageAllocation";

let store = createContextStore<StorageAllocation>();

export function getAllocation(ctx: CompilerContext, name: string) {
    let t = store.get(ctx, name);
    if (!t) {
        throw Error('Allocation for ' + name + ' not found');
    }
    return t;
}

export function getAllocations(ctx: CompilerContext) {
    return getSortedTypes(ctx).map((v) => ({ allocation: getAllocation(ctx, v.name), type: v }));
}

function getSortedTypes(ctx: CompilerContext) {
    let types = Object.values(getAllTypes(ctx)).filter((v) => v.kind === 'struct' || v.kind === 'contract');
    let structs = types.filter(t => t.kind === 'struct');
    let refs = (src: TypeDescription) => {
        let res: TypeDescription[] = []
        let t = new Set<string>();
        for (let f of src.fields) {
            let r = f.type;
            while (r.kind !== 'direct') {
                r = r.inner;
            }

            let tp = getType(ctx, r.name);
            if (tp.kind === 'struct') {
                if (!t.has(tp.name)) {
                    t.add(r.name);
                    res.push(tp);
                }
            }
        }
        return res;
    }
    structs = topologicalSort(structs, refs);
    structs = [...structs, ...types.filter((v) => v.kind === 'contract')];
    return structs;
}

function resolveFieldSize(ctx: CompilerContext, src: FieldDescription, type: TypeRef = src.type): { bits: number, refs: number } {

    // Optional type
    if (src.type.kind === 'optional') {
        let r = resolveFieldSize(ctx, src, src.type);
        return { bits: r.bits + 1, refs: r.refs };
    }

    // Direct type
    let td = getType(ctx, src.type.name);
    if (td.kind === 'primitive') {
        if (td.name === 'Int') {
            return { bits: 257, refs: 0 };
        } else if (td.name === 'Bool') {
            return { bits: 1, refs: 0 };
        } else if (td.name === 'Cell') {
            return { bits: 0, refs: 1 }
        } else {
            throw Error('Unknown primitive type: ' + td.name);
        }
    } else if (td.kind === 'struct') {
        let allocation = getAllocation(ctx, src.type.name);
        return allocation.root.size;
    } else if (td.kind === 'contract') {
        throw Error('Contract type not supported for storage');
    }

    // Unknown type
    throw Error('Unknown field type: ' + src.type.kind);
}

function allocateField(ctx: CompilerContext, src: FieldDescription, type: TypeRef, optional: boolean = false): StorageField {
    if (src.type.kind === 'optional') {
        return allocateField(ctx, src, src.type.inner, true);
    }

    let td = getType(ctx, src.type.name);
    if (td.kind === 'primitive') {
        if (td.name === 'Int') {
            return { index: src.index, size: { bits: 257, refs: 0 }, name: src.name, kind: optional ? 'int-optional' : 'int', type: td };
        } else if (src.type.name === 'Bool') {
            return { index: src.index, size: { bits: 1, refs: 0 }, name: src.name, kind: optional ? 'int-optional' : 'int', type: td };
        } else {
            throw Error('Unknown primitive type: ' + src.type.name);
        }
    } else if (td.kind === 'struct') {
        let allocation = getAllocation(ctx, src.type.name);
        return { index: src.index, size: allocation.root.size, name: src.name, kind: optional ? 'struct-optional' : 'struct', type: td };
    } else if (td.kind === 'contract') {
        throw Error('Contract type not supported for storage');
    }
    throw Error('Unknown field type: ' + src.type.kind);
}

function allocateFields(ctx: CompilerContext, src: FieldDescription[], bits: number, refs: number): StorageCell {

    let fields: StorageField[] = [];
    let next: StorageCell | null = null;
    let used: { bits: number, refs: number } = { bits: 0, refs: 0 };

    while (src.length > 0) {
        let f = src.shift()!;
        let d = resolveFieldSize(ctx, f);
        if (d.bits > bits || d.refs > refs) {
            used.refs += 1;
            next = allocateFields(ctx, [f, ...src], 1023, 3);
            break;
        } else {
            bits -= d.bits;
            refs -= d.refs;
            used.bits += d.bits;
            used.refs += d.refs;
            fields.push(allocateField(ctx, f, f.type));
        }
    }

    return { fields, next, size: used };
}

export function resolveAllocations(ctx: CompilerContext) {

    // Load topological order of structs and contracts
    let types = getSortedTypes(ctx);

    // Generate allocations
    for (let s of types) {
        let root = allocateFields(ctx, [...s.fields], 1023, 3);
        let allocation: StorageAllocation = { prefix: null, root };
        ctx = store.set(ctx, s.name, allocation);
    }

    // Generate function allocations
    for (let t of types) {
        if (t.kind === 'contract') {
            for (let f of t.functions) {
                if (f.isPublic) {
                    let fields = f.args.map((v, i) => ({ index: i, name: v.name, type: v.type }));
                    let prefix = crc32(Buffer.from(f.name));
                    let root = allocateFields(ctx, fields, 1023 - 8 * 4, 3);
                    let allocation: StorageAllocation = { prefix, root };
                    ctx = store.set(ctx, t.name + '$$' + f.name, allocation);
                }
            }
        }
    }

    return ctx;
}