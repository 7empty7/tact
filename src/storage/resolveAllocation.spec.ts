import fs from 'fs';
import { CompilerContext } from '../ast/context';
import { resolveExpressionTypes } from '../types/resolveExpressionType';
import { getAllTypes, resolveTypeDescriptors } from '../types/resolveTypeDescriptors';
import { getAllocations, resolveAllocations } from './resolveAllocation';

const stdlib = fs.readFileSync(__dirname + '/../../stdlib/stdlib.tact', 'utf-8');
const src = `

struct Point3 {
    var a: Point;
    var b: Point2;
}

struct Point {
    var x: Int;
    var y: Int;
}

struct Point2 {
    var z: Point;
}
`;

describe('resolveAllocation', () => {
    it('should write program', () => {
        let ctx = CompilerContext.fromSources([stdlib, src]);
        ctx = resolveTypeDescriptors(ctx);
        ctx = resolveExpressionTypes(ctx);
        ctx = resolveAllocations(ctx);
        expect(getAllocations(ctx)).toMatchSnapshot();
    });
});