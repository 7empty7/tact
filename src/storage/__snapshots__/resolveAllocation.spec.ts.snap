// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`resolveAllocation should write program 1`] = `
[
  {
    "allocation": {
      "fields": [
        {
          "as": null,
          "default": true,
          "index": 0,
          "name": "bounce",
          "ref": ASTRef {},
          "type": {
            "kind": "ref",
            "name": "Bool",
            "optional": false,
          },
        },
        {
          "as": null,
          "default": undefined,
          "index": 1,
          "name": "to",
          "ref": ASTRef {},
          "type": {
            "kind": "ref",
            "name": "Address",
            "optional": false,
          },
        },
        {
          "as": null,
          "default": undefined,
          "index": 2,
          "name": "value",
          "ref": ASTRef {},
          "type": {
            "kind": "ref",
            "name": "Int",
            "optional": false,
          },
        },
        {
          "as": null,
          "default": 0n,
          "index": 3,
          "name": "mode",
          "ref": ASTRef {},
          "type": {
            "kind": "ref",
            "name": "Int",
            "optional": false,
          },
        },
        {
          "as": null,
          "default": null,
          "index": 4,
          "name": "body",
          "ref": ASTRef {},
          "type": {
            "kind": "ref",
            "name": "Cell",
            "optional": true,
          },
        },
      ],
      "prefix": null,
      "root": {
        "fields": [
          {
            "bits": 1,
            "index": 0,
            "kind": "int",
            "name": "bounce",
            "size": {
              "bits": 1,
              "refs": 0,
            },
          },
          {
            "index": 1,
            "kind": "address",
            "name": "to",
            "size": {
              "bits": 267,
              "refs": 0,
            },
          },
          {
            "bits": 257,
            "index": 2,
            "kind": "int",
            "name": "value",
            "size": {
              "bits": 257,
              "refs": 0,
            },
          },
          {
            "bits": 257,
            "index": 3,
            "kind": "int",
            "name": "mode",
            "size": {
              "bits": 257,
              "refs": 0,
            },
          },
          {
            "index": 4,
            "inner": {
              "index": 4,
              "kind": "cell",
              "name": "body",
              "size": {
                "bits": 0,
                "refs": 1,
              },
            },
            "kind": "optional",
            "name": "body",
            "size": {
              "bits": 1,
              "refs": 1,
            },
          },
        ],
        "next": null,
        "size": {
          "bits": 783,
          "refs": 1,
        },
      },
    },
    "type": {
      "ast": {
        "fields": [
          {
            "as": null,
            "id": 166,
            "init": true,
            "kind": "def_field",
            "name": "bounce",
            "ref": ASTRef {},
            "type": {
              "id": 164,
              "kind": "type_ref_simple",
              "name": "Bool",
              "optional": false,
              "ref": ASTRef {},
            },
          },
          {
            "as": null,
            "id": 168,
            "init": undefined,
            "kind": "def_field",
            "name": "to",
            "ref": ASTRef {},
            "type": {
              "id": 167,
              "kind": "type_ref_simple",
              "name": "Address",
              "optional": false,
              "ref": ASTRef {},
            },
          },
          {
            "as": null,
            "id": 170,
            "init": undefined,
            "kind": "def_field",
            "name": "value",
            "ref": ASTRef {},
            "type": {
              "id": 169,
              "kind": "type_ref_simple",
              "name": "Int",
              "optional": false,
              "ref": ASTRef {},
            },
          },
          {
            "as": null,
            "id": 173,
            "init": 0n,
            "kind": "def_field",
            "name": "mode",
            "ref": ASTRef {},
            "type": {
              "id": 171,
              "kind": "type_ref_simple",
              "name": "Int",
              "optional": false,
              "ref": ASTRef {},
            },
          },
          {
            "as": null,
            "id": 176,
            "init": null,
            "kind": "def_field",
            "name": "body",
            "ref": ASTRef {},
            "type": {
              "id": 174,
              "kind": "type_ref_simple",
              "name": "Cell",
              "optional": true,
              "ref": ASTRef {},
            },
          },
        ],
        "id": 177,
        "kind": "def_struct",
        "message": false,
        "name": "SendParameters",
        "ref": ASTRef {},
      },
      "fields": [
        {
          "as": null,
          "default": true,
          "index": 0,
          "name": "bounce",
          "ref": ASTRef {},
          "type": {
            "kind": "ref",
            "name": "Bool",
            "optional": false,
          },
        },
        {
          "as": null,
          "default": undefined,
          "index": 1,
          "name": "to",
          "ref": ASTRef {},
          "type": {
            "kind": "ref",
            "name": "Address",
            "optional": false,
          },
        },
        {
          "as": null,
          "default": undefined,
          "index": 2,
          "name": "value",
          "ref": ASTRef {},
          "type": {
            "kind": "ref",
            "name": "Int",
            "optional": false,
          },
        },
        {
          "as": null,
          "default": 0n,
          "index": 3,
          "name": "mode",
          "ref": ASTRef {},
          "type": {
            "kind": "ref",
            "name": "Int",
            "optional": false,
          },
        },
        {
          "as": null,
          "default": null,
          "index": 4,
          "name": "body",
          "ref": ASTRef {},
          "type": {
            "kind": "ref",
            "name": "Cell",
            "optional": true,
          },
        },
      ],
      "functions": {},
      "init": null,
      "kind": "struct",
      "name": "SendParameters",
      "receivers": {},
    },
  },
  {
    "allocation": {
      "fields": [
        {
          "as": null,
          "default": undefined,
          "index": 0,
          "name": "x",
          "ref": ASTRef {},
          "type": {
            "kind": "ref",
            "name": "Int",
            "optional": false,
          },
        },
        {
          "as": null,
          "default": undefined,
          "index": 1,
          "name": "y",
          "ref": ASTRef {},
          "type": {
            "kind": "ref",
            "name": "Int",
            "optional": false,
          },
        },
      ],
      "prefix": null,
      "root": {
        "fields": [
          {
            "bits": 257,
            "index": 0,
            "kind": "int",
            "name": "x",
            "size": {
              "bits": 257,
              "refs": 0,
            },
          },
          {
            "bits": 257,
            "index": 1,
            "kind": "int",
            "name": "y",
            "size": {
              "bits": 257,
              "refs": 0,
            },
          },
        ],
        "next": null,
        "size": {
          "bits": 514,
          "refs": 0,
        },
      },
    },
    "type": {
      "ast": {
        "fields": [
          {
            "as": null,
            "id": 305,
            "init": undefined,
            "kind": "def_field",
            "name": "x",
            "ref": ASTRef {},
            "type": {
              "id": 304,
              "kind": "type_ref_simple",
              "name": "Int",
              "optional": false,
              "ref": ASTRef {},
            },
          },
          {
            "as": null,
            "id": 307,
            "init": undefined,
            "kind": "def_field",
            "name": "y",
            "ref": ASTRef {},
            "type": {
              "id": 306,
              "kind": "type_ref_simple",
              "name": "Int",
              "optional": false,
              "ref": ASTRef {},
            },
          },
        ],
        "id": 308,
        "kind": "def_struct",
        "message": false,
        "name": "Point",
        "ref": ASTRef {},
      },
      "fields": [
        {
          "as": null,
          "default": undefined,
          "index": 0,
          "name": "x",
          "ref": ASTRef {},
          "type": {
            "kind": "ref",
            "name": "Int",
            "optional": false,
          },
        },
        {
          "as": null,
          "default": undefined,
          "index": 1,
          "name": "y",
          "ref": ASTRef {},
          "type": {
            "kind": "ref",
            "name": "Int",
            "optional": false,
          },
        },
      ],
      "functions": {},
      "init": null,
      "kind": "struct",
      "name": "Point",
      "receivers": {},
    },
  },
  {
    "allocation": {
      "fields": [
        {
          "as": null,
          "default": undefined,
          "index": 0,
          "name": "z",
          "ref": ASTRef {},
          "type": {
            "kind": "ref",
            "name": "Point",
            "optional": false,
          },
        },
      ],
      "prefix": null,
      "root": {
        "fields": [
          {
            "index": 0,
            "kind": "struct",
            "name": "z",
            "size": {
              "bits": 514,
              "refs": 0,
            },
            "type": {
              "ast": {
                "fields": [
                  {
                    "as": null,
                    "id": 305,
                    "init": undefined,
                    "kind": "def_field",
                    "name": "x",
                    "ref": ASTRef {},
                    "type": {
                      "id": 304,
                      "kind": "type_ref_simple",
                      "name": "Int",
                      "optional": false,
                      "ref": ASTRef {},
                    },
                  },
                  {
                    "as": null,
                    "id": 307,
                    "init": undefined,
                    "kind": "def_field",
                    "name": "y",
                    "ref": ASTRef {},
                    "type": {
                      "id": 306,
                      "kind": "type_ref_simple",
                      "name": "Int",
                      "optional": false,
                      "ref": ASTRef {},
                    },
                  },
                ],
                "id": 308,
                "kind": "def_struct",
                "message": false,
                "name": "Point",
                "ref": ASTRef {},
              },
              "fields": [
                {
                  "as": null,
                  "default": undefined,
                  "index": 0,
                  "name": "x",
                  "ref": ASTRef {},
                  "type": {
                    "kind": "ref",
                    "name": "Int",
                    "optional": false,
                  },
                },
                {
                  "as": null,
                  "default": undefined,
                  "index": 1,
                  "name": "y",
                  "ref": ASTRef {},
                  "type": {
                    "kind": "ref",
                    "name": "Int",
                    "optional": false,
                  },
                },
              ],
              "functions": {},
              "init": null,
              "kind": "struct",
              "name": "Point",
              "receivers": {},
            },
          },
        ],
        "next": null,
        "size": {
          "bits": 514,
          "refs": 0,
        },
      },
    },
    "type": {
      "ast": {
        "fields": [
          {
            "as": null,
            "id": 310,
            "init": undefined,
            "kind": "def_field",
            "name": "z",
            "ref": ASTRef {},
            "type": {
              "id": 309,
              "kind": "type_ref_simple",
              "name": "Point",
              "optional": false,
              "ref": ASTRef {},
            },
          },
        ],
        "id": 311,
        "kind": "def_struct",
        "message": false,
        "name": "Point2",
        "ref": ASTRef {},
      },
      "fields": [
        {
          "as": null,
          "default": undefined,
          "index": 0,
          "name": "z",
          "ref": ASTRef {},
          "type": {
            "kind": "ref",
            "name": "Point",
            "optional": false,
          },
        },
      ],
      "functions": {},
      "init": null,
      "kind": "struct",
      "name": "Point2",
      "receivers": {},
    },
  },
  {
    "allocation": {
      "fields": [
        {
          "as": null,
          "default": undefined,
          "index": 0,
          "name": "a",
          "ref": ASTRef {},
          "type": {
            "kind": "ref",
            "name": "Point",
            "optional": false,
          },
        },
        {
          "as": null,
          "default": undefined,
          "index": 1,
          "name": "b",
          "ref": ASTRef {},
          "type": {
            "kind": "ref",
            "name": "Point2",
            "optional": false,
          },
        },
      ],
      "prefix": null,
      "root": {
        "fields": [
          {
            "index": 0,
            "kind": "struct",
            "name": "a",
            "size": {
              "bits": 514,
              "refs": 0,
            },
            "type": {
              "ast": {
                "fields": [
                  {
                    "as": null,
                    "id": 305,
                    "init": undefined,
                    "kind": "def_field",
                    "name": "x",
                    "ref": ASTRef {},
                    "type": {
                      "id": 304,
                      "kind": "type_ref_simple",
                      "name": "Int",
                      "optional": false,
                      "ref": ASTRef {},
                    },
                  },
                  {
                    "as": null,
                    "id": 307,
                    "init": undefined,
                    "kind": "def_field",
                    "name": "y",
                    "ref": ASTRef {},
                    "type": {
                      "id": 306,
                      "kind": "type_ref_simple",
                      "name": "Int",
                      "optional": false,
                      "ref": ASTRef {},
                    },
                  },
                ],
                "id": 308,
                "kind": "def_struct",
                "message": false,
                "name": "Point",
                "ref": ASTRef {},
              },
              "fields": [
                {
                  "as": null,
                  "default": undefined,
                  "index": 0,
                  "name": "x",
                  "ref": ASTRef {},
                  "type": {
                    "kind": "ref",
                    "name": "Int",
                    "optional": false,
                  },
                },
                {
                  "as": null,
                  "default": undefined,
                  "index": 1,
                  "name": "y",
                  "ref": ASTRef {},
                  "type": {
                    "kind": "ref",
                    "name": "Int",
                    "optional": false,
                  },
                },
              ],
              "functions": {},
              "init": null,
              "kind": "struct",
              "name": "Point",
              "receivers": {},
            },
          },
        ],
        "next": {
          "fields": [
            {
              "index": 1,
              "kind": "struct",
              "name": "b",
              "size": {
                "bits": 514,
                "refs": 0,
              },
              "type": {
                "ast": {
                  "fields": [
                    {
                      "as": null,
                      "id": 310,
                      "init": undefined,
                      "kind": "def_field",
                      "name": "z",
                      "ref": ASTRef {},
                      "type": {
                        "id": 309,
                        "kind": "type_ref_simple",
                        "name": "Point",
                        "optional": false,
                        "ref": ASTRef {},
                      },
                    },
                  ],
                  "id": 311,
                  "kind": "def_struct",
                  "message": false,
                  "name": "Point2",
                  "ref": ASTRef {},
                },
                "fields": [
                  {
                    "as": null,
                    "default": undefined,
                    "index": 0,
                    "name": "z",
                    "ref": ASTRef {},
                    "type": {
                      "kind": "ref",
                      "name": "Point",
                      "optional": false,
                    },
                  },
                ],
                "functions": {},
                "init": null,
                "kind": "struct",
                "name": "Point2",
                "receivers": {},
              },
            },
          ],
          "next": null,
          "size": {
            "bits": 514,
            "refs": 0,
          },
        },
        "size": {
          "bits": 514,
          "refs": 1,
        },
      },
    },
    "type": {
      "ast": {
        "fields": [
          {
            "as": null,
            "id": 300,
            "init": undefined,
            "kind": "def_field",
            "name": "a",
            "ref": ASTRef {},
            "type": {
              "id": 299,
              "kind": "type_ref_simple",
              "name": "Point",
              "optional": false,
              "ref": ASTRef {},
            },
          },
          {
            "as": null,
            "id": 302,
            "init": undefined,
            "kind": "def_field",
            "name": "b",
            "ref": ASTRef {},
            "type": {
              "id": 301,
              "kind": "type_ref_simple",
              "name": "Point2",
              "optional": false,
              "ref": ASTRef {},
            },
          },
        ],
        "id": 303,
        "kind": "def_struct",
        "message": false,
        "name": "Point3",
        "ref": ASTRef {},
      },
      "fields": [
        {
          "as": null,
          "default": undefined,
          "index": 0,
          "name": "a",
          "ref": ASTRef {},
          "type": {
            "kind": "ref",
            "name": "Point",
            "optional": false,
          },
        },
        {
          "as": null,
          "default": undefined,
          "index": 1,
          "name": "b",
          "ref": ASTRef {},
          "type": {
            "kind": "ref",
            "name": "Point2",
            "optional": false,
          },
        },
      ],
      "functions": {},
      "init": null,
      "kind": "struct",
      "name": "Point3",
      "receivers": {},
    },
  },
  {
    "allocation": {
      "fields": [],
      "prefix": null,
      "root": {
        "fields": [],
        "next": null,
        "size": {
          "bits": 0,
          "refs": 0,
        },
      },
    },
    "type": {
      "ast": {
        "declarations": [
          {
            "args": [
              {
                "id": 313,
                "kind": "def_argument",
                "name": "a",
                "ref": ASTRef {},
                "type": {
                  "id": 312,
                  "kind": "type_ref_simple",
                  "name": "Int",
                  "optional": false,
                  "ref": ASTRef {},
                },
              },
              {
                "id": 315,
                "kind": "def_argument",
                "name": "b",
                "ref": ASTRef {},
                "type": {
                  "id": 314,
                  "kind": "type_ref_simple",
                  "name": "Int",
                  "optional": false,
                  "ref": ASTRef {},
                },
              },
            ],
            "attributes": [
              {
                "ref": ASTRef {},
                "type": "public",
              },
            ],
            "id": 316,
            "kind": "def_function",
            "name": "main",
            "ref": ASTRef {},
            "return": null,
            "statements": [],
          },
        ],
        "id": 317,
        "kind": "def_contract",
        "name": "Sample",
        "ref": ASTRef {},
      },
      "fields": [],
      "functions": {
        "main": {
          "args": [
            {
              "as": null,
              "name": "a",
              "ref": ASTRef {},
              "type": {
                "kind": "ref",
                "name": "Int",
                "optional": false,
              },
            },
            {
              "as": null,
              "name": "b",
              "ref": ASTRef {},
              "type": {
                "kind": "ref",
                "name": "Int",
                "optional": false,
              },
            },
          ],
          "ast": {
            "args": [
              {
                "id": 313,
                "kind": "def_argument",
                "name": "a",
                "ref": ASTRef {},
                "type": {
                  "id": 312,
                  "kind": "type_ref_simple",
                  "name": "Int",
                  "optional": false,
                  "ref": ASTRef {},
                },
              },
              {
                "id": 315,
                "kind": "def_argument",
                "name": "b",
                "ref": ASTRef {},
                "type": {
                  "id": 314,
                  "kind": "type_ref_simple",
                  "name": "Int",
                  "optional": false,
                  "ref": ASTRef {},
                },
              },
            ],
            "attributes": [
              {
                "ref": ASTRef {},
                "type": "public",
              },
            ],
            "id": 316,
            "kind": "def_function",
            "name": "main",
            "ref": ASTRef {},
            "return": null,
            "statements": [],
          },
          "isGetter": false,
          "isMutating": true,
          "isPublic": true,
          "name": "main",
          "returns": {
            "kind": "void",
          },
          "self": "Sample",
        },
      },
      "init": null,
      "kind": "contract",
      "name": "Sample",
      "receivers": {},
    },
  },
]
`;
